Ghost RC coefficients near postprocessing boundaries

GiudGiud opened the pull request:
This could also turn in a general interface for this family of functors. 

Realistically, I do not have time to make it.

Closes https://github.com/idaholab/moose/issues/21991

GiudGiud committed:
Add routines to get boundary elements and neighbor elements from a mesh
Fix some typos in coupleable interface
refs #21991'

GiudGiud committed:
Rework the boundary neighbor elems retrieval for only active elements
Naturally skip boundaries that are not local to the process instead of erroring
refs #21991

GiudGiud committed:
Make volumetric flow rate postprocessor force the UO to ghost some data near boudnaries they care about
refs #21991

GiudGiud committed:
Allow a change of name in the density functor material for the density functor, to be able to look at the density at various pressures
refs #16809

GiudGiud committed:
Ghost a data for elements near a boundary and a process' domain boundary, refs #21991

GiudGiud committed:
Only add the gravity term for the directions where gravity is non-null, refs #16283

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/1169805/) on 6a3b2a8 wanted to post the following:

Your code requires style changes.  

A patch was auto generated and copied [here](https://mooseframework.inl.gov/docs/PRs/22078/clang_format/style.patch)  
You can directly apply the patch by running, in the top level of your repository:

`curl -s https://mooseframework.inl.gov/docs/PRs/22078/clang_format/style.patch | git apply -v`  

Alternatively, with your repository up to date and in the top level of your repository:

`git clang-format fea94ab7070512a22ea5e115c3d092f63bacb8b5`

moosebuild commented:
Job [Documentation](https://civet.inl.gov/job/1169863/) on 1a59ceb wanted to post the following:

View the site [here](https://mooseframework.inl.gov/docs/PRs/22078/site)

This comment will be updated on new commits.

lindsayad commented:
Tag me when you want me to take a look at this

GiudGiud committed:
Add test for refined boundary
uniform refinement is used as a replacement for only refining a single block because refineBlockGenerator doesnt work for distributed meshes. We care about distributed meshes. The results should not vary an inch when we switch to partial refinement so this should be ok
see #22087 for RefineBlockGen issue

GiudGiud commented:
Assuming it passes the new test in all configurations, I'd like a review

GiudGiud commented:
should be good for realz
(only error was an unused variable)

moosebuild commented:
Job [Coverage](https://civet.inl.gov/job/1175705/) on fef1799 wanted to post the following:


## Framework coverage

<table>
  <thead>
    <tr>
      <th></th>
      <th>b79b92</th>
      <th colspan=4>#22078 fef179</th>
    </tr>
    <tr>
      <th></th>
      <th>Total</th>
      <th>Total</th>
      <th>+/-</th>
      <th>New</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <td><b>Rate</b></td>
      <td>84.17%</td>
      <td>84.15%</td>
      <td>-0.02%</td>
      <td>25.00%</td>
    </tr>
    <tr>
      <td><b>Hits</b></td>
      <td>79112</td>
      <td>79112</td>
      <td>-</td>
      <td>6</td>
    </tr>
    <tr>
      <td><b>Misses</b></td>
      <td>14875</td>
      <td>14896</td>
      <td>+21</td>
      <td>18</td>
    </tr>
  </tbody>
</table>

[Diff coverage report](https://mooseframework.inl.gov/docs/PRs/22078/coverage_diff/framework)

[Full coverage report](https://mooseframework.inl.gov/docs/PRs/22078/coverage/framework)


## Modules coverage

### Navier stokes

<table>
  <thead>
    <tr>
      <th></th>
      <th>b79b92</th>
      <th colspan=4>#22078 fef179</th>
    </tr>
    <tr>
      <th></th>
      <th>Total</th>
      <th>Total</th>
      <th>+/-</th>
      <th>New</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <td><b>Rate</b></td>
      <td>85.44%</td>
      <td>85.44%</td>
      <td>+0.01%</td>
      <td>89.47%</td>
    </tr>
    <tr>
      <td><b>Hits</b></td>
      <td>9745</td>
      <td>9768</td>
      <td>+23</td>
      <td>34</td>
    </tr>
    <tr>
      <td><b>Misses</b></td>
      <td>1661</td>
      <td>1664</td>
      <td>+3</td>
      <td>4</td>
    </tr>
  </tbody>
</table>

[Diff coverage report](https://mooseframework.inl.gov/docs/PRs/22078/coverage_diff/navier_stokes)

[Full coverage report](https://mooseframework.inl.gov/docs/PRs/22078/coverage/navier_stokes)

## Full coverage reports

<details>
<summary>Reports</summary>

- [`framework`](https://mooseframework.inl.gov/docs/PRs/22078/coverage/framework)
- [`chemical_reactions`](https://mooseframework.inl.gov/docs/PRs/22078/coverage/chemical_reactions)
- [`combined`](https://mooseframework.inl.gov/docs/PRs/22078/coverage/combined)
- [`contact`](https://mooseframework.inl.gov/docs/PRs/22078/coverage/contact)
- [`electromagnetics`](https://mooseframework.inl.gov/docs/PRs/22078/coverage/electromagnetics)
- [`external_petsc_solver`](https://mooseframework.inl.gov/docs/PRs/22078/coverage/external_petsc_solver)
- [`fluid_properties`](https://mooseframework.inl.gov/docs/PRs/22078/coverage/fluid_properties)
- [`fsi`](https://mooseframework.inl.gov/docs/PRs/22078/coverage/fsi)
- [`functional_expansion_tools`](https://mooseframework.inl.gov/docs/PRs/22078/coverage/functional_expansion_tools)
- [`geochemistry`](https://mooseframework.inl.gov/docs/PRs/22078/coverage/geochemistry)
- [`heat_conduction`](https://mooseframework.inl.gov/docs/PRs/22078/coverage/heat_conduction)
- [`level_set`](https://mooseframework.inl.gov/docs/PRs/22078/coverage/level_set)
- [`misc`](https://mooseframework.inl.gov/docs/PRs/22078/coverage/misc)
- [`navier_stokes`](https://mooseframework.inl.gov/docs/PRs/22078/coverage/navier_stokes)
- [`peridynamics`](https://mooseframework.inl.gov/docs/PRs/22078/coverage/peridynamics)
- [`phase_field`](https://mooseframework.inl.gov/docs/PRs/22078/coverage/phase_field)
- [`porous_flow`](https://mooseframework.inl.gov/docs/PRs/22078/coverage/porous_flow)
- [`ray_tracing`](https://mooseframework.inl.gov/docs/PRs/22078/coverage/ray_tracing)
- [`rdg`](https://mooseframework.inl.gov/docs/PRs/22078/coverage/rdg)
- [`reactor`](https://mooseframework.inl.gov/docs/PRs/22078/coverage/reactor)
- [`richards`](https://mooseframework.inl.gov/docs/PRs/22078/coverage/richards)
- [`solid_properties`](https://mooseframework.inl.gov/docs/PRs/22078/coverage/solid_properties)
- [`stochastic_tools`](https://mooseframework.inl.gov/docs/PRs/22078/coverage/stochastic_tools)
- [`tensor_mechanics`](https://mooseframework.inl.gov/docs/PRs/22078/coverage/tensor_mechanics)
- [`thermal_hydraulics`](https://mooseframework.inl.gov/docs/PRs/22078/coverage/thermal_hydraulics)
- [`xfem`](https://mooseframework.inl.gov/docs/PRs/22078/coverage/xfem)

</details>

## Warnings

- `framework` new line coverage rate 25.00% is less than the suggested 90.0%
- `navier_stokes` new line coverage rate 89.47% is less than the suggested 90.0%



This comment will be updated on new commits.

GiudGiud commented:
@lindsayad

lindsayad commented on framework/include/mesh/MooseMesh.h:
```diff
@@ -418,11 +418,34 @@ class MooseMesh : public MooseObject, public Restartable, public PerfGraphInterf
+  getBoundariesToActiveLocalElemIds() const;
+
+  /**
+   * Return all ids of elements which have a side which is part of a boundary/sideset.
+   * Note that boundaries are sided.
+   * @param bid the id of the boundary/sideset of interest
```
As this is developer documentation, I think we can be 100% explicit that this is only about sidesets (nodesets be gone!)
```suggestion
   * Return all ids of elements which have a side which is part of a sideset.
   * @param bid the id of the sideset of interest
```

GiudGiud replied:
forgot to click I ll fix it later

GiudGiud replied:
ok coaught it

lindsayad commented on framework/include/mesh/MooseMesh.h:
```diff
@@ -418,11 +418,34 @@ class MooseMesh : public MooseObject, public Restartable, public PerfGraphInterf
+  const std::unordered_set<dof_id_type> getBoundaryActiveLocalElemIds(BoundaryID bid) const;
+
+  /**
+   * Return all ids of neighbors of elements which have a side which is part of a boundary/sideset.
```
```suggestion
   * Return all ids of neighbors of elements which have a side which is part of a sideset.
```

lindsayad commented on framework/include/mesh/MooseMesh.h:
```diff
@@ -418,11 +418,34 @@ class MooseMesh : public MooseObject, public Restartable, public PerfGraphInterf
+   * Return all ids of neighbors of elements which have a side which is part of a boundary/sideset.
+   * Note that boundaries are sided, this is on the neighbor side. For the sideset side, use
+   * getBoundariesActiveLocalElemIds.
+   * Note that while the element is local and active, the neighbor is not guaranteed to be local,
+   * it could be ghosted.
```
```suggestion
   * Note that while the element is local and active, the neighbor is not guaranteed to be local or active
```

lindsayad replied:
what happens if you don't have any side neighbor layers ghosted like for CFEM and you call this method when there are non-local neighbors?

GiudGiud replied:
you should not get them with this code, because you ll retrieve some remote_elems and we exclude them

lindsayad replied:
:+1: Might want to mention that in the doxygen

GiudGiud replied:
done

lindsayad commented on framework/include/mesh/MooseMesh.h:
```diff
@@ -418,11 +418,34 @@ class MooseMesh : public MooseObject, public Restartable, public PerfGraphInterf
+   * Note that boundaries are sided.
+   * @param bid the id of the boundary/sideset of interest
+   */
+  const std::unordered_set<dof_id_type> getBoundaryActiveLocalElemIds(BoundaryID bid) const;
```
```suggestion
  std::unordered_set<dof_id_type> getBoundaryActiveLocalElemIds(BoundaryID bid) const;
```

lindsayad commented on framework/include/mesh/MooseMesh.h:
```diff
@@ -418,11 +418,34 @@ class MooseMesh : public MooseObject, public Restartable, public PerfGraphInterf
+   * it could be ghosted.
+   * @param bid the id of the boundary/sideset of interest
+   */
+  const std::unordered_set<dof_id_type> getBoundaryActiveNeighborElemIds(BoundaryID bid) const;
```
```suggestion
  std::unordered_set<dof_id_type> getBoundaryActiveNeighborElemIds(BoundaryID bid) const;
```

lindsayad commented on framework/src/mesh/MooseMesh.C:
```diff
@@ -1000,10 +1000,57 @@ MooseMesh::getBoundaryElementRange()
+{
+  // Vector of boundary elems is updated every mesh update
+  std::unordered_set<dof_id_type> neighbor_elems;
+  for (const auto & local_elem : _bnd_elems)
```
I think we should be able to use a structured binding for the element, side, and boundary ID here. I believe this matches Case 3 here: https://en.cppreference.com/w/cpp/language/structured_binding

GiudGiud replied:
ok I tried

lindsayad reviewed (changes requested):
partway through

lindsayad commented on modules/navier_stokes/src/materials/RhoFromPTFunctorMaterial.C:
```diff
@@ -22,23 +22,28 @@ RhoFromPTFunctorMaterial::validParams()
 {
-  addFunctorProperty<ADReal>(NS::density,
+  addFunctorProperty<ADReal>(_density_name,
                              [this](const auto & r, const auto & t) -> ADReal
                              { return _fluid.rho_from_p_T(_pressure(r, t), _temperature(r, t)); });
   addFunctorProperty<ADReal>(
-      NS::time_deriv(NS::density),
+      NS::time_deriv(_density_name),
```
This works, but you could also do `"density_name"` and it would get worked out during the call to `addFunctorProperty`

GiudGiud replied:
yeah that s better

lindsayad commented on modules/navier_stokes/src/userobjects/INSFVRhieChowInterpolator.C:
```diff
@@ -439,6 +448,34 @@ INSFVRhieChowInterpolator::finalize()
+  for (auto elem_id : _moose_mesh.getBoundaryActiveLocalElemIds(boundary_id))
+  {
+    const auto & elem = _moose_mesh.elemPtr(elem_id);
+    if (elem->processor_id() != this->processor_id())
+      for (const auto i : make_range(_dim))
+        // Adding to the a coefficient will make sure the final result gets communicated
+        addToA(elem, i, 0);
```
You should just insert into `_elements_to_push_pull` directly; then you don't loop over every dim and it's more clear what we're doing here. Alternatively you could create a new protected/private method that both this method and `addToA` can call that does thread safe insertion into `_elements_to_push_pull`.

GiudGiud replied:
you need to resize _a as well.
I ll check but I think I recall this is needed

lindsayad replied:
Yea thinking about this more, I know there are find assertions in `finalize` that would break if you didn't use `addToA` here. So let's keep the call to `addToA` but you only need to do it for component 0 to make sure we create the map entry

lindsayad commented on modules/navier_stokes/src/userobjects/INSFVRhieChowInterpolator.C:
```diff
@@ -439,6 +448,34 @@ INSFVRhieChowInterpolator::finalize()
+  for (auto elem_id : _moose_mesh.getBoundaryActiveLocalElemIds(boundary_id))
+  {
+    const auto & elem = _moose_mesh.elemPtr(elem_id);
+    if (elem->processor_id() != this->processor_id())
```
You should assert that these processor IDs are equal

GiudGiud replied:
but they are not?

GiudGiud replied:
yeah this line was not doing anything, might as well delete

GiudGiud replied:
well actually. getBoundaryActiveLocalElemIds actually gives you local + ghosted
so it can happen

I think we cant have the assert here

GiudGiud replied:
We need a name for locally owned + ghosted. 
Or should we just say local = local owned + ghosted

lindsayad replied:
The API name is totally wrong if it returns local + ghosted. Local is established to mean "owned by this processor" in libMesh and in almost all framework code. I would say all but clearly there is an aberration here

GiudGiud replied:
it does.
The code works as expected though, the name is the problem imo

GiudGiud replied:
names changed

lindsayad commented on modules/navier_stokes/src/userobjects/INSFVRhieChowInterpolator.C:
```diff
@@ -439,6 +448,34 @@ INSFVRhieChowInterpolator::finalize()
+  {
+    const auto & neighbor = _moose_mesh.queryElemPtr(neighbor_id);
+    if (neighbor->processor_id() != this->processor_id())
+      for (const auto i : make_range(_dim))
+        // Adding to the a coefficient will make sure the final result gets communicated
+        addToA(neighbor, i, 0);
```
per above discussion, let's just do this for component 0

GiudGiud replied:
done

lindsayad reviewed (changes requested):
Very nice API addition and use

GiudGiud committed:
Apply suggestions from code review
- use structured binding to clean up boundary neighbor code
- only resize a coef map once, not thrice
Co-authored-by: Alex Lindsay <alexlindsay239@gmail.com>

GiudGiud commented on modules/navier_stokes/src/materials/RhoFromPTFunctorMaterial.C:
```diff
@@ -22,23 +22,27 @@ RhoFromPTFunctorMaterial::validParams()
                              { return _fluid.rho_from_p_T(_pressure(r, t), _temperature(r, t)); });
   addFunctorProperty<ADReal>(
-      NS::time_deriv(NS::density),
+      NS::time_deriv("density_name"),
```
argh it doesnt get it right here

GiudGiud replied:
reverted.
This is fine, I doubt we d want to guess the names from the derivatives too, as that is NS-specific code at this point

lindsayad commented on framework/include/mesh/MooseMesh.h:
```diff
@@ -418,11 +418,35 @@ class MooseMesh : public MooseObject, public Restartable, public PerfGraphInterf
+  /**
+   * Returns a map of boundaries to ids of elements on the boundary.
+   */
+  const std::unordered_map<boundary_id_type, std::unordered_set<dof_id_type>> &
+  getBoundariesToActiveLocalElemIds() const;
```
Per discussion today
```suggestion
  const std::unordered_map<boundary_id_type, std::unordered_set<dof_id_type>> &
  getBoundariesToActiveSemiLocalElemIds() const;
```

lindsayad commented on framework/include/mesh/MooseMesh.h:
```diff
@@ -418,11 +418,35 @@ class MooseMesh : public MooseObject, public Restartable, public PerfGraphInterf
+   * Note that boundaries are sided.
+   * @param bid the id of the boundary/sideset of interest
+   */
+  std::unordered_set<dof_id_type> getBoundaryActiveLocalElemIds(BoundaryID bid) const;
```
```suggestion
  std::unordered_set<dof_id_type> getBoundaryActiveSemiLocalElemIds(BoundaryID bid) const;
```

lindsayad commented on framework/include/mesh/MooseMesh.h:
```diff
@@ -418,11 +418,35 @@ class MooseMesh : public MooseObject, public Restartable, public PerfGraphInterf
+  getBoundariesToActiveLocalElemIds() const;
+
+  /**
+   * Return all ids of elements which have a side which is part of a boundary/sideset.
```
```suggestion
   * Return all ids of elements which have a side which is part of a sideset.
```

GiudGiud commented on framework/src/mesh/MooseMesh.C:
```diff
@@ -1000,10 +1000,60 @@ MooseMesh::getBoundaryElementRange()
+}
+
+const std::unordered_map<boundary_id_type, std::unordered_set<dof_id_type>> &
+MooseMesh::getBoundariesToActiveLocalElemIds() const
```
```suggestion
MooseMesh::getBoundariesToActiveSemiLocalElemIds() const
```

GiudGiud commented on framework/src/mesh/MooseMesh.C:
```diff
@@ -1000,10 +1000,60 @@ MooseMesh::getBoundaryElementRange()
 }
 
+std::unordered_set<dof_id_type>
+MooseMesh::getBoundaryActiveLocalElemIds(BoundaryID bid) const
```
```suggestion
MooseMesh::getBoundaryActiveSemiLocalElemIds(BoundaryID bid) const
```

GiudGiud commented on modules/navier_stokes/src/userobjects/INSFVRhieChowInterpolator.C:
```diff
@@ -439,6 +448,32 @@ INSFVRhieChowInterpolator::finalize()
+    return;
+
+  // Ghost a for the elements on the boundary
+  for (auto elem_id : _moose_mesh.getBoundaryActiveLocalElemIds(boundary_id))
```
```suggestion
  for (auto elem_id : _moose_mesh.getBoundaryActiveSemiLocalElemIds(boundary_id))
```

GiudGiud committed:
Update API name local -> semilocal since includes ghosting

Co-authored-by: Alex Lindsay <alexlindsay239@gmail.com>

GiudGiud commented:
good now?

lindsayad reviewed (approved):
more beautiful code

GiudGiud commented:
Nice thanks for the review!
